### DISCUSSION TITLE
Async calls

### DISCUSSION THREAD

**Discussion by damip** [massalabs/massa#1990] [Smart contract engine]
## Goal

Allow a smart contract to send a message to trigger another smart contract's handler asynchronously.

It works like amazon lambda + SNS:
* https://docs.aws.amazon.com/lambda/latest/dg/with-sns-create-package.html#with-sns-example-deployment-pkg-python
* https://docs.aws.amazon.com/lambda/latest/dg/with-sns-example.html

Note that all the "coins" mentioned in this document are SCE coins.

## Message format

```json
{
    "sender_address": "524546646588545445",
    "sent_at_slot": {"period": 123455, "thread": 11},
    "target_address": "2207027382739292383",
    "target_handler": "handle_message",
    "validity_start_slot": {"period": 123456, "thread": 12},
    "validity_end_slot": {"period": 123457, "thread": 16},
    "max_gas": 12334,
    "gas_price": "124.23",
    "coins": "1111.11",
    "payload": { ... any object ... }
}
```

## How to send a message during bytecode execution

* messages are sent using an ABI: `send_message(target_addrress, target_handler, validity_start, validity_end, max_gas, gas_price, coins, payload: JSON string) -> Result<(), ABIReturnError>` . Note that payload has a config defined max_async_message_payload_size
* this ABI fails in a readonly execution context
* when called, this ABI does this:
  * it consumes `compute_gas_cost_of_message_storage(context.current_slot, validity_end_slot)` of gas in the current execution. This allows making the message emission more gas-consuming when it requires storing the message in queue for longer
  * it consumes `max_gas * gas_price + coins` coins from the sender
  * it generates a Message and stores it in an Async pool

## How is the Async pool handled

* In the Async pool,  Messages are kept sorted by `priority = max_gas * gas_price`

* when a Message is added to the Async pool:
  * if the Async pool length has exceeded config.max_async_pool_length:
    * remove the lowest-priority message and reimburse "coins + max_gas*gas_price" to the message sender

* At every slot S :
  * expired messages are deleted, and "coins + max_gas*gas_price" are credited back to the message sender
  * messages that are valid at slot S are popped in highest-to-lowest priority order until they accumulate config.max_async_gas_per_slot. For each selected message M in decreasing priority order:
    * make sure that M.target_address exists and has a method called M.target_handler with the right signature, otherwise fail the execution
    * credit target_address with M.coins
    * run the target handler function with M.payload as parameter and the context:
      * max_gas = M.max_gas
      * gas_price = M.gas_price
      * slot = S
      * call_stack = [M.target_address, M.sender_address]
  * on any failure, cancel all the effects of execution and credit M.coins back to the sender
  * if there is a block at slot S, the execution of the block happens here
 
## How to receive a message (inside the smart contract)

* define a public exported function `handle_message(message)`
* this function will be called when a message arrives with the right target_address and target_handler = "handle_message"

**Comment by adrien-zinger** [comment]
About handle_message(message):
1. We need to define a "subscribe" and "unsubscribe" module to make this realistic, how do we manage that? In case you have another solution in mind, please develop how massa knows which smartcontract to wake.

I some question about the message handling. That may be in other threads...

2. Do we measure the metering of the execution in the message handler?
3. Who is paying for the consumed gas in handle_message? It's seem obvious that it should be the subscribed smart contract but I prefer to be sure :ok_hand: 
4. Depend of the _1_. Subscribing can have a price depending on how many users are subscribed? :heavy_dollar_sign:

**Reply by damip** [reply]
1. the target smart contract address and handler function name are both written inside the message, so no need for subscription. If you want to limit who is allowed to send a message to your smart contract, you can filter on message.sender_address inside the handler so that sending a message from an unauthorized address will only result in gas loss for the sender
2. yes, it is executed the same way as with ExecuteSC. The execution context is given in the description above.
3. max_gas*gas_price+coins is already paid on message emission by the sender (see description above) and the gas part is credited to nobody (so it's burned) on execution
4. no need for subscription. If you want to add a subscription system, you can implement a list of addresses in your smart contract to which you will send messages one by one

**Reply by adrien-zinger** [reply]
:ok: thanks for all!

**Comment by yvan-sraka** [comment]
https://github.com/massalabs/sc-runtime/discussions/7

**Comment by qdrn** [comment]
Looks good to me.

Not completely related but at some point I think it would be interesting to think about a more complex mechanism for the gas price. Ideally you want to monitor autonomous SCs as little as possible. If you want to be sure that your autonomous SCs operation are included, you have to put a large gas price, even though it may currently be quite small. It would be nice to be able to put bounds on the gas price, or something like that so that if no one uses autonomous SCs it does not cost much and otherwise you can still be sure it's included.

**Comment by damip** [comment]
Protocol constants:
* max_async_message_payload_size
* max_async_pool_length
* max_async_message_end_period
* max_async_gas_per_slot
* gas prices:
  * constant gas cost for message emission
  * linear gas cost on payload length
  * linear gas cost on validity end - emission slot

**Comment by Eitu33** [comment]
After a first reflexion with Damir, one plausible implementation of the Async Pool system would be to regroup the Ledger and the Pool (of SC's async messages) in a single object called "State", which would be handled the same way the Ledger currently is. Other points where discussed such as:

The possibility of adding Events to the "State"
- Pros: everything would be handled in one single object, with a FinalState, StateChanges etc.
- Cons: it wouldn't be possible to ignore Events anymore and the State would become heavier

The possibility of handling the Async Pool the same way we currently handle Events (sending SC's messages to the Nodes, save locally for execution)
- Pros: there would be no need for a "State" object
- Cons: can't actually be done because the Async Pool will interact with Consensus, which is not the case of Events

The possibility of not saving changes for the Async Pool, but always the full Pool:
- Pros: not having to compute changes since the last final state, since messages are always added and removed when the pool reaches its limit
- Cons: cost is entirely depend of max_async_pool_length

To resume, the best implementation (at first sight) appears to be the "State" containing only the Ledger and the Async Pool, handling both of their final states and changes. Also the Pool changes could be handled by saving the added messages and removed messages after a slot, or only the added messages which would require to compute which messages were removed.
